hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/WALCellCodec.java

```diff
@@ -381,8 +385,13 @@ private static void checkLength(int len, int max) throws IOException {
     private void readCompressedValue(InputStream in, byte[] outArray, int outOffset,
         int expectedLength) throws IOException {
       int compressedLen = StreamUtils.readRawVarint32(in);
-      int read = compression.getValueCompressor().decompress(in, compressedLen, outArray,
-        outOffset, expectedLength);
+      // A partial read of the compressed bytes, depending on which compression codec is used,
+      // can cause messy IO errors. This can happen when the reader is actively tailing a file
+      // being written, for replication.
+      byte[] buffer = new byte[compressedLen];
+      IOUtils.readFully(in, buffer, 0, compressedLen);
```

Comment:
Trying to understand how this actually works, are we relying on the EOFException thrown by readFully here so that upper layers in ProtofbufReader#next() handles it?

If so, curious if the actual fix should be somewhere around ProtobufLogReader#extractHiddenEof()? I mean this works but if we extract the right EOF exception, we can avoid this copy?

Reply:
> Trying to understand how this actually works, are we relying on the EOFException thrown by readFully here so that upper layers in ProtofbufReader#next() handles it?

No.

Well, yes, that would happen, but that is not the fix.

There are two problems if we use the input stream directly:

- Sometimes the input stream lies about the number of available bytes

- If we do not read in the complete segment of the compressed stream, the decompressor, depending on type, will throw random exceptions, maybe IO exceptions, maybe others. These are not EOFExceptions. They permanently confuse the log reader. 

The fix is to always provide a complete segment of the compressed stream to the decompressor. A buffer and IOUtils.readFully is required for that.

Reply:
Thanks for the detailed comment.

> If we do not read in the complete segment of the compressed stream, the decompressor, depending on type, will throw random exceptions, maybe IO exceptions, maybe others. These are not EOFExceptions. They permanently confuse the log reader.

Ya, thought so. I think we have to live with that assumption since anyone can plugin any random codec implementation.

>  we still need to rewind both the reader and the output stream.

Where is this output stream that we need to rewind? Isn't that the job of the compression codec to clean up the state if the read() fails (or did I misunderstand something?)

What I had in my mind was an "intercepting" input stream that wraps this compressed input stream and keeps track of bytes read so far. It essentially does what IOUtils.readFully() does but without copying it into a buffer. It just throws EOFException when it is end of stream and readBytesSoFar < totalBytesToBeRead. In some sense we are intercepting the exception that the decompressor runs into even before it happens and we can be sure that it is due to this EOF and not some thing else with the above check. But then I didn't get this part of rewinding the decompressor's internal output stream.

> Flushing the writer is insufficient, although I do that too in this patch to attempt to minimize the time where a tailer might not have a complete WALedit serialization at the current end of file.

Ya, figured that, was wondering if frequent flushing is detrimental to the performance, instead we should probably harden this read path and leave flushing to the upper layers?

Reply:
> Where is this output stream that we need to rewind? Isn't that the job of the compression codec to clean up the state if the read() fails (or did I misunderstand something?)

The decompressor reads from the input stream, advancing it.

The decompressor writes to the output stream while decompressing, advancing it.

If the current decompression fails and needs to be restarted with more data, both input and output streams need to be rewound. 

When tailing a log file for replication, this problem repeats constantly. 

It is better to use IOUtils.readFully to ensure all compressed bytes for the value are available before invoking the codec.

Reply:
> What I had in my mind was an "intercepting" input stream that wraps this compressed input stream and keeps track of bytes read so far. It essentially does what IOUtils.readFully() does but without copying it into a buffer. It just throws EOFException when it is end of stream and readBytesSoFar < totalBytesToBeRead. 

Will the wrapped input stream get more bytes available unless someone read()s? 
If we have to read to get the stream to advance, is this different from IOUtils.readFully?

Reply:
No, I don't understand how you want me to do this differently. We don't control the read() in the codec implementation. We can't modify them to use some new readFully() method of an input stream. If we try to hide this with the input stream implementation itself i.e. override available() as read into a buffer to ensure availability and then feed that buffer into read() until empty, we've just complicated things for no gain, there is still a buffer that is filled to ensure availability.

Reply:
> essentially does what IOUtils.readFully() does but without copying it into a buffer.

This is the part that seems impossible. In order to do what IOUtils.readFully does you have to actually read(), and what is read in has to go *somewhere*, and it can't go to the codec until finished.

Reply:
Wait.
I just found something very stupid but effective to do, and yet valid.